### ISSUE TITLE
Automated IP Switching for Simulator/Devices

### ISSUE THREAD

**Issue by aphillipo** [facebook/react-native#4245]
I have no idea where to add this but here is how to automate (mostly, under some conditions, your milage may vary - your network connection must be en0 etc.) changing the IP address of your machine when you want to run on the device.

1) Click Your Project -> Build Phases -> Click the + -> New Run Script Phase
2) Add the following code:

    INFOPLIST="${TARGET_BUILD_DIR}/${INFOPLIST_PATH}"
    echo "writing to $INFOPLIST"
    PLISTCMD="Set :SERVER_IP $(ifconfig | grep inet\ | tail -1 | cut -d " " -f 2)"
    echo -n "$INFOPLIST" | xargs -0 /usr/libexec/PlistBuddy -c "$PLISTCMD"

Next add the following to AppDelegate.m replacing the usual jsCodeLocation line:

    NSString *jsBundleUrlString;

    if (TARGET_IPHONE_SIMULATOR) {
      jsBundleUrlString = @"http://localhost:8081/index.ios.bundle";
    } else {
      NSString *serverIP = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"SERVER_IP"];
      NSString *jsCodeUrlString = [NSString stringWithFormat:@"http://%@:8081/index.ios.bundle", serverIP];
      jsBundleUrlString = [jsCodeUrlString stringByAddingPercentEscapesUsingEncoding:NSUTF8StringEncoding];
    }

    jsCodeLocation = [NSURL URLWithString:jsBundleUrlString];

And that's it, you should now be able to switch between the simulator and a device without any thought about where you are...  Let me know if this works for you.

**Comment by jaygarcia** [comment]
This look pretty damn awesome. 

I've just been bundling with and use bundled code for devices.  I like this a lot more.

**Comment by mschwartz** [comment]
Nice work.

A couple of points/issues perhaps.

Where is TARGET_IPHONE_SIMULATOR being set?

There's no logic for production builds (e.g. without a server at all).

**Comment by jaygarcia** [comment]
TargetConditionals.h @mschwartz.

In Xcode, CMD+Click (like Webstorm, etc.)

**Comment by jaygarcia** [comment]
i've been using this with pre-bundling for phone testing.  I like @aphillipo's approach much more.
```
#if TARGET_IPHONE_SIMULATOR

  jsCodeLocation = [NSURL URLWithString:@"http://localhost:8081/index.ios.bundle?platform=ios&dev=true"];

  /**
   * OPTION 2
   * Load from pre-bundled file on disk. To re-generate the static bundle
   * from the root of your project directory, run
   *
   * $ react-native bundle --minify
   *
   * see http://facebook.github.io/react-native/docs/runningondevice.html
   */
#else

   jsCodeLocation = [[NSBundle mainBundle] URLForResource:@"main" withExtension:@"jsbundle"];

#endif

```

**Comment by mschwartz** [comment]
Yeah...

I was suggesting #if / else if / else to handle production, target, and
emulator..,

**Comment by aphillipo** [comment]
I should have said, this is for development only.  Feel free to add a version to automate the build based upon the RELEASE config stuff too.

This is more about when someone comes over to your desk and you want to show them what you;ve been working on on device.  The mad scrabble to get the right IP etc. is annoying.

**Comment by mschwartz** [comment]
Understood.

FWIW, my development machine has IP 192.168.0.14.  I just used that IP in
AppDelegate.m and it works in both simulator and on device.

There was a bug that prevented this from working when one of iOS9.1 or
Xcode 7.1 came out.  It didn't work with the 9.x simulators.  I was able to
install the 8.x simulators which did work.  It seems to be fixed in one of
the more recent (newer than 0.12) versions of RN.

Your solution is excellent if you have multiple people working on the
project or you use multiple machines (work/home or laptop/desktop kind of
thing).  In those cases, you can't rely on a fixed IP like 192.168.0.14.

Cheers

**Comment by aphillipo** [comment]
Yeah, I've just been consistently working from cafe's and hotel lobbies on a project outside of work.  It's annoying having to change it and getting out of sync so I decided to fix it!

For APIs the app uses does anyone know how I'd access the Info.plist value from the React/JS side of the app?

**Comment by mschwartz** [comment]
Try this:

ifconfig | grep inet\ | tail -1 | cut -d " " -f 2

I think that will work for any en* interface.

**Comment by aphillipo** [comment]
I've updated the original code with your suggestion, @mschwartz.

**Comment by mschwartz** [comment]
Someone needs to test that out on a system with en0 and en1, maybe with
both connected.

**Comment by skevy** [comment]
@aphillipo this is a great tip!

I'm not sure this should make it in and be default behavior, but perhaps this should exist in a "Tips & Tricks" section of the docs? Thoughts @ide?

**Comment by mschwartz** [comment]
@skevy I did this up even better.

https://medium.com/modus-create-front-end-development/automated-ip-configuration-for-react-native-development-7bbf9a7a18bc#.5nh2789rj

@aphillipo did the work that inspired me.

**Comment by mschwartz** [comment]
Same blog post here:

http://moduscreate.com/automated-ip-configuration-for-react-native-development/

fully automated, including fix for setting IP for the debugging web socket :)

**Comment by aphillipo** [comment]
Amazing stuff Michael - thanks for the credit and improving the solution considerably :+1:

**Comment by satya164** [comment]
Do you think this is something that can be in core?

**Comment by aphillipo** [comment]
I don't see why not; just needs a way to override it maybe?  A force IP address setting?  All the other stuff other than finding your IP consistently should be totally robust.

**Comment by satya164** [comment]
@aphillipo Yeah, it definitely will be.

**Comment by shoreflyer** [comment]
Just tried this on a fresh app and it works like a charm. 
http://moduscreate.com/automated-ip-configuration-for-react-native-development/

+1 for making this part of the process...

**Comment by jaygarcia** [comment]
Dope!

**Comment by jacobbubu** [comment]
I'm looking for it for long time, great job! Hope that could be merged to the core.

**Comment by skevy** [comment]
@vjeux what are your thoughts about having something like this in core (assuming somebody sends a PR that's robust enough as to not break people's envs)? Not sure if you have any internal concerns that may warrant this not being a core feature.

**Comment by vjeux** [comment]
I have no context on this and would defer to @frantic. But reading the description, it sounds like we would want something like that in core. Whatever hack we can get to make people's experience less painful is welcome.

**Comment by aphillipo** [comment]
The only part of this that can go wrong is the IP detection.  Therefore I would suggest that any pull request include a commented out way to override the detected IP. 

    // OVERRIDE_DETECTED_IP = '127.0.0.1'; 

Maybe that should be in Info.plist?

The key parts to the pull request are adding the source code and making sure the scripts are added when the project is initialised.  Then we'd need to look at changing the docs.  Probably would consist of just removing some bits and adding a comment about overriding the IP for weird networking configs.

In addition for this process to be totally complete generating a release mainJS files could also be part of the script.  All it would require is running "react-native bundle --minify" if the build is set to release.

To sum up the following conditions would now be automated:

1. Running locally (automatically selecting localhost on the simulator)
2. Running on device in DEBUG mode (automatically allowing debugging on browser, referencing the machine the JS is on over the network based on automated IP detection)
3. Building for Release on Device (Building the JS and setting mainBundle to be the file path).

Things that would need to be figured out:

1. IP Address detection fails.
2. Warnings about running on device when there is no network?
3. Building for RELEASE on the simulator (is this necessary and would it just work?)

I can have a look at this if you want a PR.

**Comment by aarongreenwald** [comment]
I really like this, it helped a lot. The only thing is that it failed for me when I had more than just en0. When I change it to `ifconfig en0 | grep inet\ | tail -1 | cut -d " " -f 2` it works regardless of whether I have ethernet connected or not. Of course I don't know how many other environments will break with that change. @mschwartz

**Comment by adamterlson** [comment]
@aphillipo "For pointing APIs at my local machine does anyone know how I'd access the Info.plist value from the React/JS side of the app?"

(+1) On this.  Perhaps a different PR/effort, but this would be immensely useful.

**Comment by mschwartz** [comment]
@aarongreenwald 

If you have more than one network device (e.g. both en0 and en1 active), then my detection will perhaps fail and choose the wrong one.  It is also likely to fail if you have something like VirtualBox installed with its bridge devices.

I suggest we can add another #if to the code:

```C
#define AUTO_DETECT_IP
//#undef AUTO_DECTECT_IP
// uncomment the #undef to use the old code

#ifdef AUTO_DETECT_IP
// ... my suggested code based on @aphillipo's work
#else
// the existing code that people can edit the IP manually
// including the method 1 and method 2 both
#endif
```

**Comment by mschwartz** [comment]
```
 ~  ifconfig -a | grep inet\
	inet 127.0.0.1 netmask 0xff000000
	inet 192.168.0.104 netmask 0xffff0000 broadcast 192.168.255.255
 ~  ifconfig -a | grep inet\  | grep -v 127.0.0.1
	inet 192.168.0.104 netmask 0xffff0000 broadcast 192.168.255.255
```

**Comment by aarongreenwald** [comment]
@mschwartz But then it's all or nothing. If you have more than one active network device you have to change your IP every day (or however often you're issued a new IP address, I think I get it daily). 

What if we parametrized the network device to look for, default to any? This way I could say I always want en0 and I only have to do it once. For people with only one network device, or where their wireless card is always last, the default will work. 

I don't know, maybe that makes the configuration too complicated and isn't worth the cost. I went ahead and fixed it to en0 for my code base anyway so it isn't pressing for me.

**Comment by mschwartz** [comment]
The issue is you take your laptop to work and use ethernet.  en1 it is?

Then you go to Starbucks and you use wifi.  en0 it is?

Or you and I work on a project together.  Your laptop uses en0 wifi, my iMac uses en1 wifi.  When either of us commits and pushes our code to the repo, we stomp on the other's changes.

Like I said, for the odd cases where there's more than one interface live and the script isn't detecting the right one, you can do the manual/painful thing we do today: uncomment the #undef, and edit the IP manually (or in your case, hardcode en0).

If you have two interfaces up, there's no way to automatically pick which one you REALLY want...

**Comment by mschwartz** [comment]
```
~  netstat -in | grep en
en0   1500  <Link#4>    ac:87:a3:0f:fe:3d     5443     0     9862     0     0
en1   1500  <Link#5>    b8:09:8a:bc:ee:8d 240392969     0 338836605     0     0
en1   1500  fe80::ba09: fe80:5::ba09:8aff 240392969     - 338836605     -     -
en1   1500  192.168.0/16  192.168.0.104   240392969     - 338836605     -     -
en2   1500  <Link#6>    0a:00:00:39:06:70        0     0        0     0     0
en3   1500  <Link#7>    0a:00:00:39:06:71        0     0        0     0     0
 ~  netstat -in | grep en | grep /
en1   1500  192.168.0/16  192.168.0.104   240393212     - 338836846     -     -
```

**Comment by mschwartz** [comment]
```
 ~  ifconfig | grep inet\ | grep -v 127.0 | head -1 | cut -d " " -f 2
192.168.0.104
 ~  ifconfig | grep inet\ | grep -v 127.0 | tail -1 | cut -d " " -f 2
192.168.0.104
```

The first variation returns the first IP address that's not 127.0.0.1

The second variation returns the last IP address that's not 127.0.0.1

**Comment by skevy** [comment]
@aphillipo go ahead and start on a PR for this. Happy to help review, provide tips and get it merged into core.

**Comment by mschwartz** [comment]
I would have done a PR, but I couldn't figure out how the react guys are
creating the template for the xcodeproj file.

Maybe some documentation I missed somewhere?  Or maybe something to add to
the main README.md for developers?

**Comment by satya164** [comment]
@mschwartz It's in here - https://github.com/facebook/react-native/tree/master/local-cli/generator-ios

**Comment by satya164** [comment]
I don't know much about this. @ide Any idea on this?

**Comment by skevy** [comment]
@mschwartz what I would do (and there may be an easier way) is `react-native init` a new project, change the XCode project file appropriately, and apply the diff to the template file.

**Comment by skevy** [comment]
(and by apply the diff, I mean copy over the build phase)

**Comment by mschwartz** [comment]
Are these hash values in the .prxproj file somehow tied to the specific project?

They look like uuid of some sort, but could be used as a key in some internal Xcode table I'm unaware of.

I'll try your suggestion anyhow.

**Comment by mschwartz** [comment]
https://github.com/facebook/react-native/pull/4681

Pull request created.

**Comment by rclai** [comment]
Guys, this is really awesome, great work. It is truly annoying having to press play on Xcode, only to be met with the red screen of death because I forgot to change the dev URL.

**Comment by kwcto** [comment]
http://stackoverflow.com/a/13322549/45852

`INFOPLIST="${TARGET_BUILD_DIR}/${INFOPLIST_PATH}"`
`echo "writing to $INFOPLIST"`
`PLISTCMD="Add :SERVER_IP string $(ifconfig | head -30 | grep inet\ | tail -1 | cut -d " " -f 2)"`
`echo -n "$INFOPLIST" | xargs -0 /usr/libexec/PlistBuddy -c "$PLISTCMD" || true`
`PLISTCMD="Set :SERVER_IP $(ifconfig | head -30 | grep inet\ | tail -1 | cut -d " " -f 2)"`
`echo -n "$INFOPLIST" | xargs -0 /usr/libexec/PlistBuddy -c "$PLISTCMD" || true`

**Comment by rturk** [comment]
This can help.
I'm using Mac so I can simply use mymacsharingname.local so jsBundleUrlString becomes @"http://mymacsharingname.local:8081/index.ios.bundle";

**Comment by aleclarson** [comment]
#6362 and #6345 both have a go at fixing this issue

**Comment by jondot** [comment]
You can now try this https://github.com/jondot/react-native-network-boot

**Comment by lacker** [comment]
Looks like this is fixed now

**Comment by aphillipo** [comment]
Great!
